Documment logging formats (#6255)
kubernetes/community: contributors/devel/sig-instrumentation/logging.md

environments may wish to run at V(3) or V(4). If you wish to change the log
level, you can pass in `-v=X` where X is the desired maximum level to log.

## Logging Formats

> Kubernetes supports multiple logging formats to allow users flexibility in picking format that best matches their needs.

Reviewer: ```suggestion
Kubernetes supports multiple logging formats to allow users flexibility in picking a format that best matches their needs.
```

> As new feature and logging formats are proposed we expect that Kubernetes community might find itself in disagreement about what formats and features should be officially supported.

Reviewer: ```suggestion
As new features and logging formats are proposed, we expect that the Kubernetes community might find itself in disagreement about what formats and features should be officially supported.
```

> This section will serve as a developer documentation describing high level goals of each format, giving a holistic vision of logging options and allow community to resolve any future disagreement when making changes.

Reviewer: ```suggestion
This section will serve as a developer documentation describing high level goals of each format, giving a holistic vision of logging options and allowing the community to resolve any future disagreement when making changes.
```

### Text logging format

* Default format in Kubernetes.
* Purpose: Maintain backward compatibility with klog format.

> * Human-readable, multiline support.

Reviewer: I think human-readable is also a purpose, not just some accidental aspect of it. I would write it like this:
```
* Purpose:
  * Maintain backward compatibility with klog format.
  * Human-readable, with support for printing multi-line strings with actual line breaks.
```

Author: I will fix it in git. For future https://github.blog/changelog/2020-02-26-multi-line-code-suggestions-beta/

	line             The line number
	msg              The user-supplied message
```

See more in [here](https://github.com/kubernetes/klog/blob/9ad246211af1ed84621ee94a26fcce0038b69cd1/klog.go#L581-L597)

### JSON logging format

* Requires passing `--logging-format=json` to enable
* Purpose: Provide structured metadata in efficient way.

> * Machine-readable, no multiline support.

Reviewer: ```suggestion
* Machine-readable, multi-line strings are quoted to fit into a single output line.
```

> * Full support for structured metadata.

Reviewer: ```suggestion
* Full support for structured metadata, using `logr.MarshalLog` when available.
```
